lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,306 @@
+      }
+      var formatCharacter = _blocks.formatCharacters[i];
+      var number = _extractNumFromDateTime(
+        formatCharacter,
+        datetime,
+      ).toString();
+      if (number.length > length) {
+        if (formatCharacter == _fractionSecondCode) {
+          //Special case, as we want fractional seconds to be the leading digits
+          number = number.substring(length);
+        } else {
+          number = number.substring(number.length - length);
+        }
+      } else if (number.length < length) {
+        number = number.padLeft(length, '0');
+      }
+      buffer.write(number);
+    }
+    if (_blocks.length > 0) {
+      var lastEnd = _blocks.ends.last;
+      if (lastEnd < pattern.length) {
+        buffer.write(pattern.substring(lastEnd, pattern.length));
+      }
+    }
+    return buffer.toString();
+  }
+
+  int _extractNumFromDateTime(int? formatChar, DateTime dateTime) {
+    switch (formatChar) {
+      case _yearCode:
+        return dateTime.year;
+      case _centuryCode:
+        return (dateTime.year / 100).floor();
+      case _decadeCode:
+        return (dateTime.year / 10).floor();
+      case _monthCode:
+        return dateTime.month;
+      case _dayCode:
+        return dateTime.day;
+      case _hourCode:
+        return dateTime.hour;
+      case _minuteCode:
+        return dateTime.minute;
+      case _secondCode:
+        return dateTime.second;
+      case _fractionSecondCode:
+        return dateTime.microsecond;
+    }
+    throw AssertionError("Unreachable, the key is checked in the constructor");
+  }
+
+  /// Parse a string [formattedDateTime] to a local [DateTime] as specified in the
+  /// [pattern], substituting missing values with a default. Throws an exception
+  /// on failure to parse.
+  DateTime decode(String formattedDateTime) {
+    return _decode(formattedDateTime, isUtc, true);
+  }
+
+  /// Same as [decode], but will not throw on parsing erros, instead using
+  /// the default value as if the format char was not present in the [pattern].
+  DateTime tryDecode(String formattedDateTime) {
+    return _decode(formattedDateTime, isUtc, false);
+  }
+
+  DateTime _decode(
+    String formattedDateTime,
+    bool isUtc,
+    bool throwOnError,
+  ) {
+    var characters = formattedDateTime.codeUnits;
+    var year = 0;
+    var century = 0;
+    var decade = 0;
+    var month = 1;
+    var day = 1;
+    var hour = 0;
+    var minute = 0;
+    var second = 0;
+    var microsecond = 0;
+    for (int i = 0; i < _blocks.length; i++) {
+      var char = _blocks.formatCharacters[i];
+      var number = _extractNumFromString(characters, i, throwOnError);
+      if (number != null) {
+        if (char == _fractionSecondCode) {
+          //Special case, as we want fractional seconds to be the leading digits
+          var numberLength = _blocks.ends[i] - _blocks.starts[i];
+          number *= _powersOfTen[6 - numberLength + 1];
```

Comment:
If the `_powersOfTen` start at `1`, you don't need the `+ 1` here.

Reply:
This does make it easier - as a mathematician, starting from `1` never entered my mind somehow ;)